api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	}
+	if req.UpdatedBefore != nil {
+		ands = append(ands, bson.M{"updated_at": bson.M{"$lt": req.UpdatedBefore.AsTime()}})
+	}
+
+	// Created after/before
+	if req.CreatedAfter != nil {
+		ands = append(ands, bson.M{"created_at": bson.M{"$gt": req.CreatedAfter.AsTime()}})
+	}
+	if req.CreatedBefore != nil {
+		ands = append(ands, bson.M{"created_at": bson.M{"$lt": req.CreatedBefore.AsTime()}})
+	}
+
+	// Apply paging info
+	comp := "$lt"
+	if req.MoreToken != 0 {
+		if req.Ascending {
+			comp = "$gt"
+		}
+		t := time.Unix(0, req.MoreToken)
+		ands = append(ands, bson.M{"created_at": bson.M{comp: &t}})
+	}
+
+	if len(ands) > 0 {
+		filter["$and"] = ands
+	}
+
+	cursor, err := s.col.Find(ctx, filter, findOpts)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "querying txns: %v", err)
+	}
+	defer cursor.Close(ctx)
+	var txns []txn
+	err = cursor.All(ctx, &txns)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding txns query results: %v", err)
+	}
+
+	more := false
+	var startAt *time.Time
+	if len(txns) > 0 {
+		lastCreatedAt := &txns[len(txns)-1].CreatedAt
+		filter["created_at"] = bson.M{comp: *lastCreatedAt}
+		res := s.col.FindOne(ctx, filter)
+		if res.Err() != nil && res.Err() != mongo.ErrNoDocuments {
+			return nil, status.Errorf(codes.Internal, "checking for more data: %v", err)
+		}
+		if res.Err() != mongo.ErrNoDocuments {
+			more = true
+			startAt = lastCreatedAt
+		}
+	}
+	var pbTxns []*pb.Txn
+	for _, rec := range txns {
+		pbTxn, err := toPbTxn(rec)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "converting txn to pb: %v", err)
+		}
+		pbTxns = append(pbTxns, pbTxn)
+	}
+	res := &pb.ListTxnsResponse{
+		Txns: pbTxns,
+		More: more,
+	}
+	if startAt != nil {
+		res.MoreToken = startAt.UnixNano()
+	}
+	return res, nil
+}
+
+func (s *Service) Summary(ctx context.Context, req *pb.SummaryRequest) (*pb.SummaryResponse, error) {
+	type entityCount struct {
+		ID    interface{} `bson:"_id"`
+		Count int64       `bson:"count"`
+	}
+	type stats struct {
+		ID    string  `bson:"_id"`
+		Total int64   `bson:"total"`
+		Avg   float64 `bson:"avg"`
+		Min   int64   `bson:"min"`
+		Max   int64   `bson:"max"`
+	}
+	type report struct {
+		All            []entityCount `bson:"all"`
+		ByMessageState []entityCount `bson:"by_message_state"`
+		Waiting        []entityCount `bson:"waiting"`
+		UniqueFrom     []entityCount `bson:"unique_from"`
+		UniqueTo       []entityCount `bson:"unique_to"`
+		SentFilStats   []stats       `bson:"sent_fil_stats"`
+	}
+
+	createdAtMatch := bson.M{}
+	if req.Before != nil {
+		createdAtMatch["$lt"] = req.Before.AsTime()
+	}
+	if req.After != nil {
+		createdAtMatch["$gt"] = req.After.AsTime()
+	}
+	match := bson.M{}
+	if len(createdAtMatch) > 0 {
+		match["created_at"] = createdAtMatch
+	}
+	cursor, err := s.col.Aggregate(ctx, bson.A{
+		bson.M{"$match": match},
+		bson.M{"$facet": bson.M{
```

Comment:
Facet lets you pass the incoming result set to any number of parallel sub pipelines, pretty cool.